inst/include/cpp11/integers.hpp

```diff
@@ -145,32 +145,22 @@ inline int na() {
   return NA_INTEGER;
 }
 
-// forward declarations
-typedef r_vector<double> doubles;
-
-template <>
-bool is_na(const double& x);
+// forward declaration
 
-template <>
-double r_vector<double>::operator[](const R_xlen_t pos) const;
+typedef r_vector<double> doubles;
 
-inline integers as_integers(sexp x) {
+inline integers as_integers(SEXP x) {
   if (TYPEOF(x) == INTSXP) {
-    return as_cpp<integers>(x);
+    return integers(x);
   } else if (TYPEOF(x) == REALSXP) {
-    doubles xn = as_cpp<doubles>(x);
-    R_xlen_t len = xn.size();
-    writable::integers ret(len);
-    for (R_xlen_t i = 0; i < len; ++i) {
-      double el = xn[i];
-      if (is_na(el)) {
-        ret[i] = na<int>();
-      } else if (is_convertible_without_loss_to_integer(el)) {
-        ret[i] = static_cast<int>(el);
-      } else {
+    doubles xn(x);
+    writable::integers ret = writable::integers(xn.size());
+    std::transform(xn.begin(), xn.end(), ret.begin(), [](double value) {
+      if (!is_convertible_without_loss_to_integer(value)) {
```

Comment:
Do we need to consider the case when have `(double)NA_INTEGER` in a `doubles`, currently the `static_cast<int>` will turn this into NA_INTEGER so we end up with NA although we started with something that is not NA (in the `double` land). 

Should `is_convertible_without_loss_to_integer()` return `false` in that case. R does this: 

```
> as.integer(cpp11::cpp_eval("(double)NA_INTEGER"))
[1] NA
Warning message:
NAs introduced by coercion to integer range 
```